ddl/ddl_api.go

```diff
@@ -2133,11 +2134,20 @@ func checkTwoRangeColumns(ctx sessionctx.Context, curr, prev *model.PartitionDef
 	return false, nil
 }
 
-func parseAndEvalBoolExpr(ctx sessionctx.Context, expr string, tbInfo *model.TableInfo) (bool, error) {
-	e, err := expression.ParseSimpleExprWithTableInfo(ctx, expr, tbInfo)
+func parseAndEvalBoolExpr(ctx sessionctx.Context, l, r string, coll string, tbInfo *model.TableInfo) (bool, error) {
```

Comment:
How about replacing `coll string` with `colInfo *model.ColumnInfo` so that you can get the charset and set https://github.com/pingcap/tidb/pull/20937/files#diff-b7c54f7dad7257cbe4fa04b8d76c9bed694e2e2e7c5599f45eb2172d8d4ff8a7R2150 with:

```
e.SetCharsetAndCollate(chs, coll)
```

Reply:
updated